### PR TITLE
feat(while-let): add Rust-style `while let PATTERN = SCRUTINEE { ... }`

### PR DESCRIPTION
**Pull request by codeshaunted**
## Summary

Adds Rust-style **`while let PATTERN = SCRUTINEE { BODY }`** to BAML.

It's a refutable loop header: each iteration re-evaluates the scrutinee and re-tests a (possibly refutable) pattern, runs the body while it matches, and exits on the first non-match — mirroring Rust. Bindings introduced by the pattern are scoped to the body and re-bound each iteration. It produces unit and supports `break`/`continue`. An **irrefutable** pattern warns (`E0137`) — like `if let` — rather than being rejected like a plain `let`.

```baml
function drain(stack: int[]) -> int {
    let total = 0;
    while let [let head, ..] = stack {
        total = total + head;
        let _ = stack.shift();
    }
    total
}
```

## Design

First-class statement (`Stmt::WhileLet`), not a desugaring — BAML has no surface `loop`/`match` to route through, and a first-class node keeps diagnostics, LSP, and formatting faithful (the same call `for` and `if let` already make). It lowers directly in MIR to the existing loop + pattern-test machinery, so there are **no VM / bytecode / emit changes**.

Touched layers:
- **Parser:** `while` + one-token `let` lookahead → `WHILE_LET_STMT` (mirrors the `if` / `if let` split).
- **AST:** `Stmt::WhileLet { pattern, scrutinee, body }` + CST→AST lowering.
- **HIR:** one block scope per statement with the pattern bindings.
- **TIR:** refutable pattern lowering via the `if let` path (warn-not-reject if irrefutable), per-iteration `snapshot`/`restore` of scoped locals, `divergence = false` (the body may run zero times); forward-reference and throw-fact collection.
- **MIR:** a loop header that re-tests the pattern each iteration via `lower_pattern_test`; per-iteration **fresh binding cells** (so a closure made each pass captures a distinct cell); `break`/`continue` via `LoopContext`; watched-local cleanup. The scrutinee is materialized once per header pass — a bare-path scrutinee resolves to its own local so body mutations are observed on the next test.
- **Diagnostics:** `E0137` (irrefutable `while let`) + LSP mapping.
- **Formatter / CFG visualization / onionskin / debug renderers.**

## Testing

- **Parser** unit tests (`while let` parses to `WHILE_LET_STMT`; scrutinee destructure-suppression; top-level array pattern; plain `while` unaffected).
- **Formatter** idempotency tests (no parens, no `else`).
- **17 engine integration tests** (`bex_engine/tests/while_let.rs`): drain / re-evaluation, `break`, `continue` re-tests the pattern, class & nested destructure, fresh-cell closure capture, zero-iteration fall-through, non-divergence with a reachable post-loop `throw`, outer-mutation-persists / inner-shadow-restored, body-local no-leak, scrutinee-evaluated-once-per-iteration, nested-loop `break` isolation, array-rest drain, union-typed binding, `throw` propagate/catch, and `return` from the body.
- **Diagnostic snapshot** for the irrefutable-pattern warning.

Full suite green locally.

## Note

Depends on the union-typed-binding fix (#3642, already merged to canary) — a `while let x: A | B = …` head exercises exactly that path.

## Summary by CodeRabbit

* **New Features**
  * Introduced while-let loops: pattern-driven conditional iteration that repeats while the pattern matches and stops on mismatch.

* **Diagnostics**
  * New warning for irrefutable patterns in while-let that always match.

* **Formatting**
  * Formatter support and tests to round-trip while-let syntax without changing unrelated while forms.

* **Tests**
  * Extensive unit and end-to-end tests covering semantics, scoping, control flow, and edge cases.

* **Tooling & Rendering**
  * IDE diagnostics, visualizations, and internal renderers updated to display while-let loops.

### CODE REVIEW

File: baml_language/crates/baml_fmt/src/ast/statements.rs [hunk 1]
```diff
@@ -413,6 +418,94 @@ impl Printable for WhileStmt {
...
+        printer.print_raw_token(&self.keyword);
+        printer.print_str(" ");
+        // `while let PATTERN = SCRUTINEE { ... }` — the pattern carries its own
+        // leading `let`; no parens around the pattern or scrutinee (mirrors
+        // `if let`, unlike plain `while` which canonicalises parens).
+        printer.print(&self.pattern, shape.clone());
+        printer.print_str(" ");
+        printer.print_raw_token(&self.equals);
+        printer.print_str(" ");
+        printer.print(&*self.scrutinee, shape.clone());
```
[the remark this answers is not in the document]
Comment (codeshaunted) [reply baml_language/crates/baml_fmt/src/ast/statements.rs:506]: Good catch — fixed in ea8e86da3. Added an optional `let_keyword: Option<t::Let>` to `WhileLetStmt`, consumed via `next_if_kind(KW_LET)` and printed before the pattern (mirrors `LetStmt`), so `while let [x] = xs { ... }` round-trips through the formatter. Added a regression test (`test_while_let_array_pattern`).
